### DISCUSSION TITLE
How to test an htmx application?

### DISCUSSION THREAD

**Discussion by guettli** [bigskysoftware/htmx#680] [Q&A]
There is a [question on Stackoverflow](https://stackoverflow.com/questions/67133382/unit-testing-htmx) about how to test htmx  based applications.

I think there should be a small section about this in the docs.

**Comment by BoPeng** [comment]
That is framework dependent, for Django/Python, I use [`django-htmx`](https://github.com/adamchainz/django-htmx) to process htmx request to be processed by Django, and I do something like the following to create a fake request with HTMX header for testing:

```python
from django_htmx.middleware import HtmxDetails

def htmx_request(user=None,
                     get_url=None,
                     post_url=None,
                     post_data={},
                     trigger=None,
                     **kwargs):
    headers = dict(HTTP_HX_Request="true", HTTP_HX_Trigger_Name=trigger)
    if get_url is not None:
        request = RequestFactory().get(
            get_url, {x: y for x, y in kwargs.items() if y is not None}, **headers)
    if post_url is not None:
        kwargs.update(headers)
        request = RequestFactory().post(post_url, data=post_data, **kwargs)
    request.htmx = HtmxDetails(request)
    request.user = user if user else AnonymousUser()
    return request
```

**Comment by guettli** [comment]
Every professional framework has at least one chapter in his docs about how to write tests.

I think there are three ways:

- use the test tools of you backend framework to test your backend.
- e2e tests with Playwright  (or similar tool)
- maybe some JS based testing which does not need a full browser. But I  have not done this yet.

This is just a gentle reminder that a chapter in the docs would be useful.

**Comment by kensmith** [comment]
Seconded. I just came here to ask this. I had been thinking of using Selenium/Playwright/Puppeteer but had a discussion with someone recently where they had a bad experience with e2e due to fragile tests and just pushed all testing to the unit test level. This vibes with my intuition. My backend is in Go so it's likely possible to do something similar to what @BoPeng  does in Python but how are people testing their HTMX applications in general?

**Reply by BoPeng** [reply]
The test we have done stayed at the request/view level with the synthetic htmx requests. We spent a huge amount of time on selenium-based tests but these tests were very slow and many of them would fail randomly despite our best efforts to robustify them. We eventually gave up on them because those failures blocked our code review/merge pipelines.

**Comment by Telroshan** [comment]
I personally use Playwright that was already mentioned above, but there's indeed no "offical" way to test htmx-based apps for now.

Btw see https://github.com/bigskysoftware/htmx/discussions/2360 for a related discussion on race-conditions and how to make reliable htmx-based tests.

If someone feels like it, PRs improving documentation are always welcome!

**Comment by DevOpsCraftsman** [comment]
I end up with a strategy which is… a little bit weird, but which works.

But there is one important pre-condition:

The template engine syntax you’re using should have an homologue in the JS world.

On this specific project, which is in Kotlin / Spring Boot, we are using Pebble.

It’s a "jinja-like" SSR template, like the ones of Django (Python) and twig (in PHP / Symfony).

Happens there is a similar one for JS, which is called nunjucks
It can renders template in both the front-end and the backend (node env, and JSDOM env)

The trick is simply to have nunjucks render the pebble (or jinja, or twig...) templates in a JSDOM environnement, and let the JS run naturally, and run assertion like you would do with any other frontend.

That’s enough for tools like Alpine

For htmx - which fires AJAX requests, asking the server for html fragments to update the current DOM - you need an additionnal trick:

MSW

With it, you can intercept those request, and render the appropriate partial templates, with nunjucks

https://www.linkedin.com/posts/the-awareness-provider_you-should-re-architect-to-make-your-front-end-activity-7428457876949512193-UwjJ

**Reply by DevOpsCraftsman** [reply]
<img width="1906" height="1784" alt="image" src="https://github.com/user-attachments/assets/3d279d44-3324-4a8e-8581-50a011f621eb" />

You will likely need a little adapter for differences in syntax

**Comment by kevinccbsg** [comment]
When I tried to add tests I started using Playwright. Actually it is great for end-to-end testing, but for day-to-day development it often felt like too much overhead, and I feel testing like a separate thing from my development.

I ended up experimenting with running UI tests directly in the browser during development, with mocked responses where needed. It made the feedback loop much faster for me.

I wrote about the workflow here in case it's useful to anyone exploring different approaches: https://dev.to/kevinccbsg/testing-htmx-apps-in-the-real-browser-with-twd-3j4e

> I'm one of the maintainers of the tool used in the article, so I'm obviously biased. I'd genuinely love to hear what people building production HTMX apps think. Does this kind of workflow solve a real problem for you, or do you prefer keeping everything in Playwright?
